ZOOKEEPER-2247

rakeshadr opened the pull request:
Created this PR using proposed "ZOOKEEPER-2247-15.patch" in the jira.

fpj commented on src/java/main/org/apache/zookeeper/server/ZooKeeperServer.java:
```diff
@@ -486,19 +474,27 @@ protected void startSessionTracker() {
         ((SessionTrackerImpl)sessionTracker).start();
     }
 
+    void setState(State state) {
```
I'd really appreciate if we could have a fair amount of comments for these new methods explaining why they are here and how they are used. The methods themselves are trivial, but the whole logic of stoping servers due to errors and such isn't necessarily trivial.

fpj commented on src/java/main/org/apache/zookeeper/server/quorum/Leader.java:
```diff
@@ -1376,4 +1382,8 @@ public static String getPacketType(int packetType) {
     }
+
+    private boolean isRunning() {
+        return self.isRunning() && zk.isRunning();
```
Is it really logical AND here or should we be using logical OR? I'm wondering if there is any chance that just one of those `isRunning` calls evaluates to true.

rakeshadr replied:
zk.isRunning() will return false when ZooKeeperServerListener notifies an error. AND is required here, because whoever is using self.isRunning() earlier(before applying this patch) should stop/exit their execution by seeing an error in the server. Does this makes sense to you?

fpj replied:
Ok, so this is currently only used in line 588, so the correct behavior is to shut down in the case either is not running, yes?

rakeshadr replied:
Yes, correct.

fpj replied:
Ok.

fpj commented on src/java/main/org/apache/zookeeper/server/ZooKeeperServerListenerImpl.java:
```diff
@@ -0,0 +1,43 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * Default listener implementation, which will be used to notify internal
```
Please expand this explanation.

rakeshadr replied:
Done, please review it again. Thanks!

fpj commented on src/java/main/org/apache/zookeeper/server/ZooKeeperServerMain.java:
```diff
@@ -146,14 +146,26 @@ public void runFromConfig(ServerConfig config) throws IOException, AdminServerEx
 
+            // Watch status of ZooKeeper server. If there is an internal error
+            // then will do a graceful shutdown.
+            while (zkServer.isRunning()) {
```
This loop works, but I'm not finding it very elegant. In particular, I don't like having a loop + sleep like this to block the thread, so I'm wondering if we can do better than this and have a call that really blocks until the condition is met. For example, we can have a latch that waits until we count down or interrupt it. What do you think?

rakeshadr replied:
For quorum it was easy because ZK has Leader/Follower/Observer thread to read the packets and has an existing polling mechanism. But for a standalone server we don't have such polling mechanism and I think need to introduce some mechanism to listen ZKServer state changes and notify the main thread. Earlier we discussed an approach by adding an extra HealthMonitorThread but decided to move that logic inside main thread - https://issues.apache.org/jira/browse/ZOOKEEPER-2247?focusedCommentId=15126934&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15126934

Thanks for the latch suggestion. But I think here also we should have a separate notifier thread to poll the server state changes and countdown latch, right?

fpj replied:
Agreed about the `HealthMonitorThread`, I'd rather not have that. The behavior we are trying to implement is this. ZooKeeper main needs to block until the `isRunning` predicate becomes false. To do it, either we poll as we are doing now in this patch and keep waking up every now and then, or we implement a blocking call in `ZooKeeperServer` that puts all callers to wait until they are notified. We can notify all when `setState` is called and the condition is met. Does it work?

rakeshadr replied:
OK. How about exposing an interface like below by ZooKeeperServer.java class and ZooKeeperServerMain will implement this using CDL logic. This implementation will be passed as constructor arg in ZooKeeperServer.java and invoked during state changes.

/**
 * Observer interface which will be used to notify the state changes.
 */
public interface ZooKeeperServerStateListener {

    /**
     * This will be invoked when the server transition to a new server state.
     *
     * @param state new server state
     */
    void stateChanged(State state);
}

ZooKeeperServerMain.java
    private static class ZooKeeperServerStateListenerImpl
            implements ZooKeeperServerStateListener {
        private CountDownLatch healthMonitorLatch;

        ZooKeeperServerStateListenerImpl(CountDownLatch healthMonitorLatch) {
            this.healthMonitorLatch = healthMonitorLatch;
        }

        @Override
        public void stateChanged(State state) {
            if (state != State.RUNNING) {
                healthMonitorLatch.countDown();
            }
        }
    }

fpj replied:
@rakeshadr The idea is fine, but I'm wondering why we would do an interface rather than just implementing directly, it is not that we envision multiple implementations of this interface, right? But correct me if I'm missing something.

rakeshadr replied:
If someone needs a different usage in standalone server, they can do a different impl of this interface and pass it through ZKServer constructor rather than extending the impl class. I thought exposing an interface would be more cleaner.

fpj replied:
@rakeshadr sure, I understand that an interface would allow us to do a different version, but do we right now envision any other way of implementing this? if not, then I'd just **not** do an interface and know that we can do one if it ever becomes a need. I'd rather avoid a proliferation of unnecessary interfaces, otherwise we will end up doing everything with an interface.

rakeshadr replied:
Agreed, I understand it. Instead of an interface I will directly make a ZooKeeperServerStateListener class and register like zks#registerStateListener(ZooKeeperServerStateListener zksStateListener);

OK?

fpj replied:
sounds good, thanks!

rakeshadr replied:
Done. Please review it again. Thanks!

fpj commented on src/java/test/org/apache/zookeeper/test/NonRecoverableErrorTest.java:
```diff
@@ -0,0 +1,178 @@
+    /**
+     * Test case for https://issues.apache.org/jira/browse/ZOOKEEPER-2247.
+     * Test to verify that even after non recoverable error (error while
+     * writing transaction log) on ZooKeeper service will be available.
```
Rephrase: Test to verify that even after non recoverable error (error while writing transaction log), ZooKeeper is still available.

rakeshadr replied:
Done. Thanks for the correcting it.

fpj commented on src/java/test/org/apache/zookeeper/test/NonRecoverableErrorTest.java:
```diff
@@ -0,0 +1,178 @@
+            // do create operation, so that injected IOException is thrown
+            zk.create(uniqueNode(), data.getBytes(), Ids.OPEN_ACL_UNSAFE,
+                    CreateMode.PERSISTENT);
+            fail("IOException is expected as error is injected in transaction log commit funtionality");
```
Rephrase: IOException is expected due to error injected to transaction log commit.

rakeshadr replied:
Done. Thanks for the correcting it.

fpj commented on src/java/test/org/apache/zookeeper/test/NonRecoverableErrorTest.java:
```diff
@@ -0,0 +1,178 @@
+        leader.getActiveServer().setZKDatabase(originalZKDatabase);
+
+        // verify that now ZooKeeper service is up and running
+        QuorumPeer currentleader = getLeaderQuorumPeer(mt);
```
We can reuse the leader variable here.

rakeshadr replied:
Done

fpj commented on src/java/test/org/apache/zookeeper/test/NonRecoverableErrorTest.java:
```diff
@@ -0,0 +1,178 @@
+        return null;
+    }
+
+    private String uniqueNode() {
```
Maybe call it uniqueZnode instead.

rakeshadr replied:
Done

fpj commented on src/java/test/org/apache/zookeeper/server/ZooKeeperServerMainTest.java:
```diff
@@ -133,6 +138,62 @@ public void shutdown() {
+     * writing transaction log) on ZooKeeper service will be available
+     */
+    @Test(timeout = 30000)
+    public void testNonRecoverableError() throws Exception {
```
Do we really need this test case given that we have the one in `NonRecoverableErrorTest.java`? This test seems to be contained in the other further down.

rakeshadr replied:
ZooKeeperServerMainTest.java test class to verify standalone cases. The newly added test is to verify the behavior in standalone server. So this test case is needed for us.

NonRecoverableErrorTest.java verifies quorum behavior. I will add class level javadoc in this class to make it more clear, is that ok?

fpj replied:
Yes, makes sense.

fpj commented on src/java/main/org/apache/zookeeper/server/quorum/Leader.java:
```diff
@@ -584,6 +584,12 @@ void lead() throws IOException, InterruptedException {
 
+                    // check leader running status
+                    if (!this.isRunning()) {
+                        shutdown("Unexpected internal error");
```
One question about this. Say we hit it here and print this message, how does the admin determines what was the actual cause of the error? Will the admin have to look for the exception that caused the problem? I'm just wondering about the troubleshooting path.

rakeshadr replied:
Yes, ZooKeeperServerListenerImpl will do error logging before setting server running status to ERROR and from that admin can determine the cause. I will change the log priority to "error" in ZooKeeperServerListenerImpl.java class. Is that OK?

LOG.error("Thread {} exits, error code {}", threadName, exitCode);

fpj replied:
Ok.

fpj commented:
I've left one new comment, and there are a bunch of requests for improvements open. Could you take care of those, please @rakeshadr?

rakeshadr commented:
Thanks @fpj. Sure, I will address all the comments in my next patch.

rakeshadr commented:
Thanks @fpj , I have tried addressing the comments, kindly take a look at the new commit again. I will  rebase the patch in the latest code base once I get a +1 for the changes.

fpj commented on src/java/main/org/apache/zookeeper/server/ZooKeeperServerMain.java:
```diff
@@ -146,14 +151,19 @@ public void runFromConfig(ServerConfig config) throws IOException, AdminServerEx
 
+            // Watch status of ZooKeeper server. It will do a graceful shutdown
+            // if the server is not running or hits an internal error.
+            healthMonitorLatch.await();
```
Small point, but I'd rather call it "shutdownLatch" or "stopLatch" rather than "healthMonitorLatch". It is not only about the health of the server, the server might be completely healthy and you still want to cross the latch to stop it, right?

rakeshadr replied:
OK. used 'shutdownLatch'

fpj commented:
@rakeshadr other than the small comment I've left, I think this is good, please upload a patch to the jira so that we can have QA running and such.

rakeshadr commented:
Thank you again for the detailed reviews. Attached patch to the jira.

fpj commented on src/java/main/org/apache/zookeeper/server/ZooKeeperServer.java:
```diff
@@ -486,19 +475,57 @@ protected void startSessionTracker() {
+    void setState(State state) {
+        this.state = state;
+        // Notify server state changes to the registered shutdown handler, if any.
+        if (zkShutdownHandler != null) {
```
@rakeshadr is there a potential race here between setting the shutdown handler and checking if it is null?

rakeshadr replied:
server#registerServerShutdownHandler() is called before server startup invocation. server#setState api will be invoked during server#startup, serverListener#notifyStopping, server#shutdown. In all these cases, the handler would have been already registered and not null(as we know this handler is only used in standalone mode). In that case, I couldn't see any race condition. Please let me know if I missed any case. Thanks!

fpj replied:
Ok, makes sense, but in this case, the null check, assuming if we ever have a regression that makes zkShutdownHandler null at this point for standalone mode, is simply allowing us to proceed without even a warning. Should we have an assertion here or at least log an error?

rakeshadr replied:
I agree to add an assert. But in case of quorum, 'shutdownHandler' will be null, so it will make noise in quorum. So I'm planning to override #setState() api in subclassess -QuorumZooKeeperServer, ReadOnlyZooKeeperServer.

    @Override
    protected void setState(State state) {
        this.state = state;
    }

For that, some additional changes are required and those are:
(1) Make ZooKeeperServer#setState() api visibility to 'protected'. Now its default, package visibility.
(2) Make ZooKeeperServer's "enum State" visibility to 'protected'

OK?

fpj replied:
@rakeshadr it sounds good to add the log error for standalone and not for the other cases. You're right that it will add noise to the log if we simply add it here.

I'm not entirely sure why you want to make `setState` protected, though. Don't we need to call it from `notifyStopping`? I'm also not sure why we would make `enum State` protected.

rakeshadr replied:
Yes, zkServer.setState() will be called inside #notifyStopping(). I'm changing #setState to protected because org.apache.zookeeper.server.quorum.QuorumZooKeeperServer and org.apache.zookeeper.server.ZooKeeperServer classes are in different package. Presently, #setState() is having default package visibility so for overriding #setState in subclass, it needs to increase the visibility, right? Same reason for "enum State" protected.

fpj replied:
@rakeshadr makes sense for `setState()`, but why do you need to override `enum State`, can't use the one you already in `ZooKeeperServer`?

rakeshadr replied:
@fpj, The purpose of increasing the visibility is not for overriding the "enum State", that is to use "State" inside QuorumZooKeeperServer.java file. While overriding "#setState()" method it needs to pass "State" as an argument, like below:
    @Override
    protected void setState(State state) {
        this.state = state;
    }

Since "State" is defined inside ZooKeeperServer.java file with default visibility, it can't be used outside the package.

I think, we have agreed on all other points and this is the only pending part. I think uploading changes to this PR with proposed code changes will help us in discussion, shall I?

fpj replied:
@rakeshadr sounds good.

rakeshadr replied:
@fpj, Done one more checkin by fixing (1) comma after errors and before then (2) #setState() api changes (3) Fixed one Raul's comment about removing extra "()" bracket. 

Please review it again. I will upload patch to jira once I get a +1 from you. Thanks!

fpj commented on src/java/main/org/apache/zookeeper/server/ZooKeeperServerListenerImpl.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Default listener implementation, which will be used to notify internal
+ * errors. For example, if some critical thread has stopped due to fatal errors
+ * then it will get notifications and will change the state of ZooKeeper server
```
@rakeshadr nit: missing a comma after errors and before then.

rakeshadr replied:
Agreed, I will do it.

fpj commented on src/java/main/org/apache/zookeeper/server/ZooKeeperServerListenerImpl.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @Override
+    public void notifyStopping(String threadName, int exitCode) {
```
@rakeshadr nit: I noticed that `handleException` logs an error message including a throwable. Here we log an info message with the exit code. It is not the point of this jira, so fine if we leave this for a different issue, but I'm thinking that it is better to consolidate the logging and have only a log error message here in this `notifyStopping` method. We would need to change the signature and pass a throwable, though.

rakeshadr replied:
OK, how about raise an imprv jira for this task and discuss separately?

fpj replied:
Ok, sounds good, it is not a critical change, but good to have.

rakeshadr replied:
Raised ZOOKEEPER-2501 to take care this comment. Will post a patch once ZK-2247 is in.

fpj commented:
+1, let's check this one in.

hanm commented:
@rakeshadr I think this pr can be closed because it's merged already.
